src/inmanta/server/services/compilerservice.py

```diff
@@ -226,6 +227,84 @@ async def _run_compile_stage(self, name: str, cmd: list[str], cwd: str, env: dic
                 # The process is still running, kill it
                 sub_process.kill()
 
+    async def ensure_compiler_venv(self) -> None:
+        """ "
+        Ensure we have a compiler venv in the project
+
+        Expected to be in a running stage
+
+        It is built as
+        1. a `.env` symlink to
+        2. a `.env-py3.12` versioned symlink
+        """
+        assert os.path.exists(self._project_dir)
+        project_dir = self._project_dir
+
+        # Eventual venv dir
+        venv_dir = os.path.join(project_dir, ".env")
+
+        # Versioned python dir
+        python_version = ".".join(platform.python_version_tuple()[0:2])
+        versioned_venv_dir = ".env-py" + python_version
+        versioned_venv_dir_full = os.path.join(project_dir, versioned_venv_dir)
+
+        # Determine the final action
+        # If we don't need to symlink, we return
+        do_create = False
+
+        if os.path.islink(venv_dir):
+            # We have a modern setup
+            try:
+                if os.path.samefile(venv_dir, versioned_venv_dir_full):
+                    # all good
+                    await self._info("Found existing venv")
+                    return None
+                else:
+                    # Wrong version
+                    os.unlink(venv_dir)
+                    await self._info(f"Creating new venv at {versioned_venv_dir_full}")
```

Comment:
Shouldn't we add more context to the log line indicating that the Python version of the venv is out-of-date?

Reply:
I see your point. Maybe it's better to keep it like this.